ansible/roles/test/tasks/pfc_wd/functional_test/functional_test.yml

```diff
@@ -0,0 +1,299 @@
+          - port_src='{{pfc_wd_test_port_id}}'
+          - ip_src='{{pfc_wd_rx_neighbor_addr}}'
+          - ip_dst='{{pfc_wd_test_neighbor_addr}}'
+          - wd_action='forward'
+        ptf_extra_options: "--relax --debug info --log-file /tmp/pfc_wd.PfcWdTest.{{ansible_date_time.iso8601}}.log "
+
+    - set_fact:
+        test_expect_file: "expect_pfc_wd_restore"
+
+    - name: Initialize loganalyzer
+      include: roles/test/files/tools/loganalyzer/loganalyzer_init.yml
+
+    - name: Wait for queue to recover from PFC storm
+      pause:
+        seconds: "{{(pfc_wd_penalty / 1000 * 2) | int}}"
+
+    - name: Check if logs contain message that PFC WD restored from deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_analyze.yml
+
+    - name: Check if logs contain message that PFC WD restored from deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_end.yml
+
+  rescue:
+    - name: Remove all temporary files from local host.
+      file:
+        name: "{{ tests_location }}/functional_test/{{ item }}"
+        state: absent
+      with_items: "{{ config_files }}"
+      connection: local
+      become: false
+
+  always:
+    - name: Delete forward config from {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_del_action.json
+        errors_expected: false
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+    - name: Remove all the temporary files created by the test.
+      include: roles/test/tasks/deinit_config_test.yml
+
+# Verify timers
+- name: Initialize config test
+  include: roles/test/tasks/init_config_test.yml
+
+- block:
+    - name: Apply forward config to {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_fwd_action.json
+        errors_expected: false
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+
+    - name: Enable routing from mgmt interface to localhost
+      sysctl:
+        name: net.ipv4.conf.eth0.route_localnet=1
+        value: 1
+        sysctl_set: yes
+
+      # Cannot use iptables module, because it does not support to_destination in 2.0
+    - name: Add iptables rule to forward syslog packets from mgmt interface to localhost
+      shell: "iptables -t nat -I PREROUTING -p udp -d {{ansible_eth0['ipv4']['address']}} --dport 514 -j DNAT --to-destination 127.0.0.1:514"
+
+    - set_fact:
+        detect_time_list: []
+        restore_time_list: []
+
+    - name: Calculate detection and restoration timings
+      include: roles/test/tasks/pfc_wd/functional_test/timer_test.yml
+      with_sequence: start=1 end=3
+
+    - debug:
+        var: "{{item}}"
+      with_items: 
+        - detect_time_list
+        - restore_time_list
+
+    - name: Verify that real detection time is not greater than configured X2
+      fail:
+        msg: Real detection time is greater than configured X2
+      when: "{{(detect_time_list | sum) / 3 > pfc_wd_poll_time * 2}}"
```

Comment:
Should be detection_time + polling_time? Also it should fail when the result is not within range[detection_time, detection_time + polling_time] not only greater.

Reply:
Added test for both bounds. Will change upper bound to detect + polling time in P2